apps/src/maze/mazeController.js

```diff
@@ -0,0 +1,363 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview JavaScript for Blockly's Maze application.
+ * @author fraser@google.com (Neil Fraser)
+ */
+
+const timeoutList = require('../lib/util/timeoutList');
+
+const AnimationsController = require('./animationsController');
+const MazeMap = require('./mazeMap');
+const drawMap = require('./drawMap');
+const getSubtypeForSkin = require('./mazeUtils').getSubtypeForSkin;
+const tiles = require('./tiles');
+
+module.exports = class MazeController {
+  constructor(level, skin, config) {
+    const Type = getSubtypeForSkin(config.skinId);
+    const subtype = new Type(this, config);
+
+    this.stepSpeed = 100;
+
+    this.level = level;
+    this.skin = skin;
+    this.startDirection;
+
+    this.map;
+    this.subtype = subtype;
+    this.animationsController;
+    this.executionInfo;
+    this.store;
+
+    this.pegmanD;
+    this.pegmanX;
+    this.pegmanY;
+
+    this.MAZE_HEIGHT;
+    this.MAZE_WIDTH;
+    this.PATH_WIDTH;
+    this.PEGMAN_HEIGHT;
+    this.PEGMAN_WIDTH;
+    this.PEGMAN_X_OFFSET;
+    this.PEGMAN_Y_OFFSET;
+    this.SQUARE_SIZE;
+
+    this.loadLevel_();
+  }
+
+  /**
+   * A few placeholder methods intended to be rebound
+   */
+  playAudio() {}
+  playAudioOnFailure() {}
+  loadAudio() {}
+  getTestResults() {}
+
+  rebindMethods(methods) {
+    this.playAudio = methods.playAudio || this.playAudio;
+    this.playAudioOnFailure = methods.playAudioOnFailure || this.playAudioOnFailure;
+    this.loadAudio = methods.loadAudio || this.loadAudio;
+    this.getTestResults = methods.getTestResults || this.getTestResults;
+  }
+
+  addReduxStore(store) {
+    this.store = store;
+  }
+
+  initWithSvg(svg) {
+    // Adjust outer element size.
+    svg.setAttribute('width', this.MAZE_WIDTH);
+    svg.setAttribute('height', this.MAZE_HEIGHT);
+
+    drawMap.default(svg, this.skin, this.subtype, this.map, this.SQUARE_SIZE);
+    this.animationsController = new AnimationsController(this, svg);
+  }
+
+  loadLevel_() {
+    // Load maps.
+    //
+    // "serializedMaze" is the new way of storing maps; it's a JSON array
+    // containing complex map data.
+    //
+    // "map" plus optionally "levelDirt" is the old way of storing maps;
+    // they are each arrays of a combination of strings and ints with
+    // their own complex syntax. This way is deprecated for new levels,
+    // and only exists for backwards compatibility for not-yet-updated
+    // levels.
+    if (this.level.serializedMaze) {
```

Comment:
Can we pass in `level` as an argument instead of adding it to `this`?

Reply:
probably, yes; in fact, the way it currently works is that we pass in `level`, `skin`, and `config` to the constructor despite the fact that `level` and `skin` are actually just properties on `config`.

I did it this way because that's what most closely mirrors how the current version of Maze is initialized, and I wanted this PR to be just a reorganization with as few functional changes as possible, with the intention of doing cleanups like that when things are more cleanly separated.